src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.element = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        final Node<T> prev = tail;
+        final Node<T> newNode = new Node<>(prev, value, null);
+        tail = newNode;
+        if (prev == null) {
+            head = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(index + " is invalid index");
+        }
+        if (index == 0) {
+            Node<T> next = head;
+            Node<T> node = new Node<>(null, value, next);
+            if (size == 0) {
+                tail = node;
+            }
+            head = node;
+            if (size > 0) {
+                next.prev = node;
+            }
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            Node<T> foundNode = getNodeByIndex(index).prev;
+            Node<T> nextNode = foundNode.next;
+            Node<T> node = new Node<>(foundNode, value, nextNode);
+            foundNode.next = node;
+            nextNode.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] arr = list.toArray();
+        if (arr.length == 0) {
+            return;
+        }
+        for (Object obj : arr) {
+            add((T) obj);
+        }
```

Comment:
```suggestion
        for (T obj : list) {
            add(obj);
        }
```
this check is redundant, if length is 0 we will just skip the loop